backend/app/core/rag/graph_builder.py

```
@@ -0,0 +1,128 @@
+        path = self.cache_dir / f"{safe_name}_graph.pkl"
+        with open(path, "wb") as f:
+            pickle.dump(graph, f)
+        return path
+
+    def load_graph(self, folder_path: str) -> Optional[nx.Graph]:
+        safe_name = Path(folder_path).name.replace(" ", "_")
+        path = self.cache_dir / f"{safe_name}_graph.pkl"
+        if not path.exists():
+            return None
+        with open(path, "rb") as f:
+            return pickle.load(f)
```

Ruff (0.15.12)

[error] `pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)